File: sequential/SudokuSequential.cpp
#include <array>
#include <iostream>
#include <vector>

using namespace std;

const int kSubgridSize = 3;
const int kGridSize = 9;

using Board = vector<vector<char>>;

// Builds a sample Sudoku board to demonstrate solver usage.
Board MakeSampleBoard()
{
    return {{'5', '3', '.', '.', '7', '.', '.', '.', '.'},
            {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
            {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
            {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
            {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
            {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
            {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
            {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
            {'.', '.', '.', '.', '8', '.', '.', '7', '9'}};
}

void PrintBoard(const Board &board)
{
    for (int row = 0; row < kGridSize; ++row)
    {
        if (row % kSubgridSize == 0 && row != 0)
            cout << "------+-------+------\n";

        for (int col = 0; col < kGridSize; ++col)
        {
            if (col % kSubgridSize == 0 && col != 0)
                cout << " |";

            cout << ' ' << board[row][col];
        }
        cout << '\n';
    }
}

class SudokuSolver
{
public:
    bool Solve(Board &board)
    {
        ResetState();

        board_ = &board;
        for (int row = 0; row < kGridSize; ++row)
        {
            for (int col = 0; col < kGridSize; ++col)
            {
                if (board[row][col] != '.')
                {
                    const int value = board[row][col] - '0';
                    PlaceNumber(value, row, col);
                }
            }
        }

        Backtrack(0, 0);
        return solved_;
    }

private:
    void ResetState()
    {
        for (auto &row : rows_)
            row.fill(0);
        for (auto &col : cols_)
            col.fill(0);
        for (auto &box : boxes_)
            box.fill(0);

        board_ = nullptr;
        solved_ = false;
    }

    bool CouldPlace(int value, int row, int col) const
    {
        const int boxIndex = (row / kSubgridSize) * kSubgridSize + col / kSubgridSize;
        return rows_[row][value] == 0 && cols_[col][value] == 0 && boxes_[boxIndex][value] == 0;
    }

    void PlaceNumber(int value, int row, int col)
    {
        const int boxIndex = (row / kSubgridSize) * kSubgridSize + col / kSubgridSize;
        ++rows_[row][value];
        ++cols_[col][value];
        ++boxes_[boxIndex][value];
        (*board_)[row][col] = static_cast<char>('0' + value);
    }

    void RemoveNumber(int value, int row, int col)
    {
        const int boxIndex = (row / kSubgridSize) * kSubgridSize + col / kSubgridSize;
        --rows_[row][value];
        --cols_[col][value];
        --boxes_[boxIndex][value];
        (*board_)[row][col] = '.';
    }

    void PlaceNextNumbers(int row, int col)
    {
        if (row == kGridSize - 1 && col == kGridSize - 1)
        {
            solved_ = true;
            return;
        }

        if (col == kGridSize - 1)
            Backtrack(row + 1, 0);
        else
            Backtrack(row, col + 1);
    }

    void Backtrack(int row, int col)
    {
        if (solved_)
            return;

        if ((*board_)[row][col] == '.')
        {
            for (int value = 1; value <= kGridSize; ++value)
            {
                if (CouldPlace(value, row, col))
                {
                    PlaceNumber(value, row, col);
                    PlaceNextNumbers(row, col);
                    if (!solved_)
                        RemoveNumber(value, row, col);
                }
            }
        }
        else
        {
            PlaceNextNumbers(row, col);
        }
    }

    array<array<int, kGridSize + 1>, kGridSize> rows_{};
    array<array<int, kGridSize + 1>, kGridSize> cols_{};
    array<array<int, kGridSize + 1>, kGridSize> boxes_{};
    Board *board_ = nullptr;
    bool solved_ = false;
};

int main()
{
    Board board = MakeSampleBoard();

    SudokuSolver solver;
    if (solver.Solve(board))
        PrintBoard(board);
    else
        cout << "No solution found.\n";

    return 0;
}

File: sequential/SudokuSequential.java
import java.util.Arrays;

public final class SudokuSequential {

    private static final int SUBGRID_SIZE = 3;
    private static final int GRID_SIZE = 9;

    private SudokuSequential() {
    }

    private static char[][] makeSampleBoard() {
        return new char[][] {
                { '5', '3', '.', '.', '7', '.', '.', '.', '.' },
                { '6', '.', '.', '1', '9', '5', '.', '.', '.' },
                { '.', '9', '8', '.', '.', '.', '.', '6', '.' },
                { '8', '.', '.', '.', '6', '.', '.', '.', '3' },
                { '4', '.', '.', '8', '.', '3', '.', '.', '1' },
                { '7', '.', '.', '.', '2', '.', '.', '.', '6' },
                { '.', '6', '.', '.', '.', '.', '2', '8', '.' },
                { '.', '.', '.', '4', '1', '9', '.', '.', '5' },
                { '.', '.', '.', '.', '8', '.', '.', '7', '9' }
        };
    }

    private static void printBoard(char[][] board) {
        for (int row = 0; row < GRID_SIZE; row++) {
            if (row % SUBGRID_SIZE == 0 && row != 0) {
                System.out.println("------+-------+------");
            }
            for (int col = 0; col < GRID_SIZE; col++) {
                if (col % SUBGRID_SIZE == 0 && col != 0) {
                    System.out.print(" |");
                }
                System.out.print(' ');
                System.out.print(board[row][col]);
            }
            System.out.println();
        }
    }

    private static final class SudokuSolver {
        private final int[][] rows = new int[GRID_SIZE][GRID_SIZE + 1];
        private final int[][] columns = new int[GRID_SIZE][GRID_SIZE + 1];
        private final int[][] boxes = new int[GRID_SIZE][GRID_SIZE + 1];
        private char[][] board;
        private boolean solved;

        boolean solve(char[][] board) {
            resetState();
            this.board = board;

            for (int row = 0; row < GRID_SIZE; row++) {
                for (int col = 0; col < GRID_SIZE; col++) {
                    if (board[row][col] != '.') {
                        placeNumber(board[row][col] - '0', row, col);
                    }
                }
            }

            backtrack(0, 0);
            return solved;
        }

        private void resetState() {
            for (int row = 0; row < GRID_SIZE; row++) {
                Arrays.fill(rows[row], 0);
                Arrays.fill(columns[row], 0);
                Arrays.fill(boxes[row], 0);
            }
            board = null;
            solved = false;
        }

        private boolean couldPlace(int value, int row, int col) {
            int boxIndex = (row / SUBGRID_SIZE) * SUBGRID_SIZE + col / SUBGRID_SIZE;
            return rows[row][value] == 0 && columns[col][value] == 0 && boxes[boxIndex][value] == 0;
        }

        private void placeNumber(int value, int row, int col) {
            int boxIndex = (row / SUBGRID_SIZE) * SUBGRID_SIZE + col / SUBGRID_SIZE;
            rows[row][value]++;
            columns[col][value]++;
            boxes[boxIndex][value]++;
            board[row][col] = (char) ('0' + value);
        }

        private void removeNumber(int value, int row, int col) {
            int boxIndex = (row / SUBGRID_SIZE) * SUBGRID_SIZE + col / SUBGRID_SIZE;
            rows[row][value]--;
            columns[col][value]--;
            boxes[boxIndex][value]--;
            board[row][col] = '.';
        }

        private void placeNextNumbers(int row, int col) {
            if (row == GRID_SIZE - 1 && col == GRID_SIZE - 1) {
                solved = true;
                return;
            }

            if (col == GRID_SIZE - 1) {
                backtrack(row + 1, 0);
            } else {
                backtrack(row, col + 1);
            }
        }

        private void backtrack(int row, int col) {
            if (solved) {
                return;
            }

            if (board[row][col] == '.') {
                for (int value = 1; value <= GRID_SIZE; value++) {
                    if (couldPlace(value, row, col)) {
                        placeNumber(value, row, col);
                        placeNextNumbers(row, col);
                        if (!solved) {
                            removeNumber(value, row, col);
                        }
                    }
                }
            } else {
                placeNextNumbers(row, col);
            }
        }
    }

    public static void main(String[] args) {
        char[][] board = makeSampleBoard();
        SudokuSolver solver = new SudokuSolver();
        if (solver.solve(board)) {
            printBoard(board);
        } else {
            System.out.println("No solution found.");
        }
    }
}